### ISSUE TITLE
No FormData in Node.js

### ISSUE THREAD

**Issue by milibopp** [evancz/elm-http#24]
Is it possible to have this module work in Node.js? I run my test suite using Node.js and was somewhat puzzled, when merely constructing a message with `Http.multipart` caused a `ReferenceError: FormData is not defined`.

Maybe I am missing something, but it looks like `FormData` is simply not present in node. Maybe [this package](https://github.com/form-data/form-data) could help?

**Comment by magopian** [comment]
Can you show how you run your test suite using Node.js? It might work to add a `var FormData = require('form-data');` in your Node.js script (but I doubt it).

For reference, the only reference I could find in the code of `FormData` is [here](https://github.com/evancz/elm-http/blob/906bccc382a85ea1140c360e3acbdb96e4c458be/src/Native/Http.js#L135)

**Comment by milibopp** [comment]
So long ago… I only vaguely remember the context and can't find the code any more.

**Comment by evancz** [comment]
I turned @process-bot on for this repo, so it should ask for http://sscce.org from now on.

Node is not an officially supported platform for Elm, so no libraries that use tasks are guaranteed to work there. If `elm-repl` gets the ability to run tasks, this may change.

@magopian, one thing to think about is OP fatigue. Ultimately, I need to assess all the issues on this repo in a coherent way. Any given OP is only gonna want to discuss things so much, so it makes sense to budget that to some extent. If _you_ get it, but I don't, and OP is done talking, things are not in a better place. Furthermore, long issues are often harder to figure out. So all I am saying here is that it makes sense to budget discussion.

The development process is batched. That is just a fact. So discussing more isn't really getting things done any sooner. Curating stuff can be good. If you look on other repos, we do "meta" issues that summarize a bunch of other closed issues. We have not have that done successfully by a contributor yet, but that's the model I like at least.

**Comment by magopian** [comment]
Thanks a lot for the feedback @evancz it makes total sense. I'm not used (yet) to batched work, I need to find the best way to contribute my free time in a meaningful way. If you have any additional pointers on how to do that, please let me know!

**Comment by evancz** [comment]
I always tell folks to get involved gradually. I think Richard and Aaron and Luke are all examples of slowly getting more involved until they are relatively active on @elm-lang repos.

So a reasonable progression is something like this:
- Help people out on elm-discuss and on Slack. First this means you see where most of the questions are, so you get a feel for what is important. Second, this means you start meeting other folks who have been around longer, know more technical stuff, know more cultural stuff.
- Create projects and blog about them. Ideally that is:
  1. A direct response to seeing what folks are asking about in the community. This way you can say, folks wonder about X and here is a way to think about that.
  2. Something you are personally passionate about. This way you can say, hey look at this cool thing I made. Here are some facets that are interesting.
- If you find some challenge that bugs you _and_ matches your skills, maybe try to make a package that addresses it. Learn more about API design in Elm by talking to users and other community members with more experience.
- By this time you will probably know more folks in the community, and have some amount of trust. If so, you can maybe start proposing work that requires coordination. Notice that Luke or Richard will say on elm-dev, I have been thinking about X. My needs are Y. I think Z is a good design. I will do this work. What does everyone think? All that before really "contributing". In other words, collaboration is never unilateral. It requires a group of people who know and trust each other working together. And the size of that group is not unlimited. Having 10 people or 1000 people will produce different quality work, so more is not always better.

In other words, the idea that "contributing = merges on github" does not make sense for a language or OS where collaboration is much more complex than on a single JS library or something. Hopefully that helps!

**Comment by magopian** [comment]
It does help a lot, thanks! I feel this answer should be widely accessible (as a `Contributing.md` or in the faq or on the official website?), as it's very clear of your (the core dev) expectations and needs. It's a clear path towards meaningful contribution, and as such, is very valuable. Thank you very much for taking the time to write this down!
